attalos/imgtxt_algorithms/sampledSM/train_tensorflow.py

```diff
@@ -0,0 +1,131 @@
+    hidden_layer = tf.matmul(hidden_layer_inputs, weights_hidden)
+    #hidden_layer = tf.nn.sigmoid(tf.matmul(hidden_layer_inputs, weights_hidden))
+    return tf.matmul(hidden_layer, weights_output), hidden_layer
+
+model_outputs, hidden_layer = single_layer_perceptron(hidden_layer_inputs, weights_hidden, weights_output)
+
+# The loss function
+poss_times_model = tf.mul(positive_examples, tf.log(tf.sigmoid(model_outputs)))
+neg_times_model =  tf.mul(negative_examples, tf.log(tf.sigmoid(-model_outputs)))
+
+poss_loss = tf.reduce_mean(poss_times_model)
+neg_loss = tf.reduce_mean(neg_times_model)
+
+loss = -1. * (poss_loss + neg_loss)
+
+# Optimizer
+optimizer = tf.train.GradientDescentOptimizer(learning_rate=learning_rate).minimize(loss)
+
+# Set up negative samples
+word_count = np.zeros(oh.vocab_size)
+for dataset in (train_dataset, test_dataset):
+    for item_index in dataset:
+        img, text = dataset[item_index]
+        word_count += oh.get_multiple(text)
+
+negative_sample_generator = negsamp.NegativeSampler(word_count)
+
+init_op = tf.initialize_all_variables()
+
+config = tf.ConfigProto()
+config.gpu_options.allow_growth=True
+
+saver = tf.train.Saver([weights_hidden, weights_output])
+
+with tf.Session(config=config) as sess:
+
+    # Initialize the tensors
+    sess.run(init_op)
+
+    # Training cycle
+    for epoch in range(training_epochs):
+        print("Training {}".format(epoch))
+        for batch in range(int(train_dataset.num_images / batch_size)):
+            # Get a batch
+            train_image_features, text_tags = train_dataset.get_next_batch(batch_size)
+            train_text_features = np.array(map(oh.get_multiple, text_tags))
+
+            negative_samples = negative_sample_generator.negsampv(train_text_features, negative_sample_count)
+
+            _, loss_result = sess.run(
+                [optimizer, loss],
+                feed_dict={
+                    hidden_layer_inputs: train_image_features,
+                    positive_examples: train_text_features,
+                    negative_examples: negative_samples,
+                }
+            )
+
+        print("\tLoss: {:.5}".format(loss_result))
+
+        # Test model
+        test_results = sess.run(
+            model_outputs,
+            feed_dict={
+                hidden_layer_inputs: train_image_features,
+            }
+        )
+        # Get the accuracy
+        model_predicted_indices = np.argpartition(test_results, -5)[:,-5:]
+        true_indices = np.argpartition(train_text_features, -5)[:, -5:]
+
+        total = 0
+        for i in range(len(model_predicted_indices)):
+            total += len(np.intersect1d(model_predicted_indices[i], true_indices[i])) / 5.
+
+        total /= float(len(model_predicted_indices))
+        print("\tAccuracy: {:.2%}\n".format(total))
+
+        if not epoch % 100:
+            # Save the variables to disk.
+            save_path = saver.save(sess, "./my_model", global_step=epoch)
```

Comment:
Likewise the save location and name should probably be something specified in arguments.